### PR TITLE
[Resolver] Use Molinillo for dependency resolution

### PR DESCRIPTION
**Pull request by segiddins**
Use a vendored version of Molinillo for resolution, just like Bundler does.
@indirect suggested I do this tonight, so I did! It could probably do with some refactoring and improved documentation, so I'd consider the PR a work in progress, but I wanted to let CI have a go.

### CODE REVIEW

File: test/rubygems/test_gem_resolver.rb [hunk 1]
```diff
@@ -95,81 +95,6 @@ def test_self_compose_sets_single
...
-
-    res = Gem::Resolver.new [a1]
-    res.ignore_dependencies = true
-
-    reqs = Gem::Resolver::RequirementList.new
-
-    res.requests a1, act, reqs
-
-    assert_empty reqs
-  end
```
Comment (tenderlove) [inline test/rubygems/test_gem_resolver.rb]: Presumably these were added for a reason?  Are they going to be added back, or, what was the reason for removing the tests?

**Comment by evanphx** [comment]
Nice! We need to eliminate the code that mutates the load path to load that vendored code though. Also all the classes loaded have to be under the Rubygems namespace. One reason for that is that a RG with this code would have a ton of errors loading bundler because they define the same top level constants from different files.

A rake task to create a vendored version of your code that concats all the files together then adds a header the moves into a different module would be best.

**Comment by evanphx** [comment]
Of! Somehow I missed that you did change the namespace. Just need to fix the require path stuff then.

**Comment by segiddins** [comment]
@evanphx yeah, this is the same vendoring code used in bundler, which also adds namespacing, since Molinillo is also a proper gem. I'm more than happy to also fix the load path, but I'm curious why it's necessary?

File: test/rubygems/test_gem_resolver.rb (hunk 1 above)
Comment (segiddins) [inline test/rubygems/test_gem_resolver.rb]: Plan to re-add them. Currently I'm not using the `requests` method but I can/will go back to it.

**Comment by segiddins** [comment]
I know I also need to fix up the UI and probably also `stats`.

**Comment by indirect** [comment]
Nice work! Super impressed by how quickly you managed to add this with a completely green build. :)

**Comment by segiddins** [comment]
@evanphx just so I can keep things moving, I'd really appreciate knowing what you'd like to see before you consider reviewing / merging this PR.

**Comment by indirect** [comment]
One additional point about this change: Molinillo has the most extensive tests of any Ruby dependency resolver, and regularly adds more as we hear from Bundler and CocoaPods users who have found edge cases. Sharing tests and fixes across Rubygems, Bundler, and CocoaPods gets a big :+1: from me.

**Comment by evanphx** [comment]
The only thing that gives me pause about merging this (and I really do want to merge it) is that test that was changed. I need to know why the test was changed because I assume that the change means this merge will deviate from some existing behavior.

**Comment by segiddins** [comment]
The `test_handle_conflict` spec was removed because Molinillo does conflict resolution internally, and thus the `handle_conflict` method that was being tested was removed

**Comment by segiddins** [comment]
@evanphx here's the summary of why some of the orders in the tests changed:

dependency resolution is a constraint satisfaction problem, and a relatively complex one at that. while the end goal is to have a fully satisfied dependency graph (and ideally at the highest possible version for each node), the path that the algorithm takes in the middle can vary a lot between implementations. What these changes in the specs show is that Molinillo is generally taking a different path on its way to determining that the initial set of constraints is unsatisfiable than the current resolver. It's reaching the same conclusion, just from a different angle, so to speak.

Hope that helps!

**Comment by segiddins** [comment]
@evanphx is there anything you'd like me to explain better?

**Comment by indirect** [comment]
For the record, I am strongly in favor of this change—I believe it is best for Ruby users if RubyGems and Bundler resolve the same dependencies the same way.

**Comment by indirect** [comment]
@segiddins can you update this with the latest Molinillo? Looks like master is ready for 2.5 changes, now that 2.4 has its own stable branch.

**Comment by segiddins** [comment]
This is now using version 0.3.0 of Molinillo, which I released this morning with the features necessary to support RubyGems.

**Comment by indirect** [comment]
\o/

**Comment by voxik** [comment]
So when is Bundler going to drop Molinillo?

**Comment by andremedeiros** [comment]
@voxik it's a fine resolver. Why would we drop it?

**Comment by voxik** [comment]
Since it is included in RubyGems, why would you ship separate copy? The best would be if Molinillo is subsumed into RubyGems and dies as independent project.

**Comment by andremedeiros** [comment]
It's also used in CocoaPods, so it makes sense to keep it separate and vendor it. Notice how we, for instance, vendor Thor for the CLI stuff as well.

Also, Bundler has to support older Rubygems versions, so we can't make any assumptions as to what environments people run.

**Comment by voxik** [comment]
I see, vendoring and bundling is new standard now, it had to come also to RubyGems one day :+1: I rather back out from this discussion and see if there is some chance to unbundle this for Fedora.

**Comment by andremedeiros** [comment]
I'm curious, why the unbundling requirement? What are you trying to achieve?

**Comment by simi** [comment]
@andremedeiros AFAIK there are [very strict rules](https://fedoraproject.org/wiki/Packaging:No_Bundled_Libraries) for library bundling in Fedora packaging.

**Comment by andremedeiros** [comment]
@simi I do understand, but the resolver is a core part of Bundler and isn't exposed. Essentially we can only make guarantees about the version we've vendored with a release, that's pretty much it.

**Comment by voxik** [comment]
https://fedoraproject.org/wiki/Packaging:No_Bundled_Libraries

And just for illustration, this is what happens when there is something bundled:

https://bugzilla.redhat.com/show_bug.cgi?id=1166064

i.e. look at the "Depends On" field, that many packages we must review due to one security issue (and there is slight chance that we are still missing something, since somebody renamed the file for example).

And there are even more interesting security issues, such as:

https://bugzilla.redhat.com/show_bug.cgi?id=1033460

And this is just because somebody copied some part of algorithm from somewhere, so they naturally share the same issues. So this issue hit Ruby as well as other totally unrelated libraries. If it was library, it would be fixed on one place once and forever.

**Comment by andremedeiros** [comment]
Right, I see your point.

I still maintain, however, that the resolver is a core part of bundler. Because it's internal (albeit vendored), we can only guarantee it works with the version it's been released with.

For this case, if we did have a security issue, we'd release a point upgrade, which I would argue could also be released by the distribution.

**Comment by segiddins** [comment]
@voxik I'm offended that you wish to see Molinillo "die" as a separate project -- I spent hundreds of hours building it so it would be usable for any project, it's certainly not just meant for RubyGems

**Comment by indirect** [comment]
@voxik you have been perfectly happy with Bundler and RubyGems as long as they had separate resolvers, even though they had more bugs than Molonillo. Deciding that Fedora needs to rip apart Bundler and RubyGems, destroying our ability to support specific versions, and destroying anyone's ability to be sure everything actually works, is not a good idea.

Bundler and RubyGems have each improved and maintained their resolvers. The fact that Molonillo exists doesn't change that.

**Comment by jrafanie** [comment]
> @voxik I'm offended that you wish to see Molinillo "die" as a separate project -- I spent hundreds of hours building it so it would be usable for any project, it's certainly not just meant for RubyGems

First of all I'm all for having rubygems and bundler share the same resolver in any way possible.  This PR makes us one step closer, so thanks @segiddins for what must have taken a long time.  :+1: 

I believe @voxik is very concerned as am I about vendoring another library in rubygems, especially considering how few people actually upgrade their rubygems outside of installing a new ruby cc @tenderlove.  

@voxik was preferring that rubygems consume molinillo so there's only 1 source of that molinillo code.  I think it can stand alone as long as rubygems/bundler/other library get it "somehow" that's not vendoring.

How would each rubygems/bundler/abc library upgrade molonillo due to a CVE or a major bug if they're vendoring them?  As a user, how would you even know that you're vulnerable via bundler/rubygems vendoring of a library?  How would I as a consumer of rubygems/bundler know if rubygems/bundler made any changes to 0.3.0 molonillo to get it to work vendored?

Dependencies are really hard and I appreciate all that the rubygems/bundler team does but it really concerns me to vender other libraries.

**Comment by segiddins** [comment]
RubyGems and Bundler need to vendor Molinillo because they're unable to have any dynamically resolved dependencies. Outside of these two projects, Molinillo can be consumed as a normal gem, but inside RubyGems/Bundler, it should merely be considered a part of the host project, not as a dependency.

**Comment by jrafanie** [comment]
Sorry for being concerned without offering an alternative. :cry:

It's good to understand the concerns though in case there's a solution those with other knowledge could create.

**Comment by segiddins** [comment]
It's just very frustrating to have this conversation _again_, after having it over on Bundler.

**Comment by jrafanie** [comment]
Sorry,  I'm late to the game... Many months and I'm not aware of the bundler discussion nor do you need to re-explain.  I appreciate all that goes into rubygems and bundler especially since many contributions are done in people's free time.  Thanks.  :heart:

**Comment by andremedeiros** [comment]
> RubyGems and Bundler need to vendor Molinillo because they're unable to have any dynamically resolved dependencies. Outside of these two projects, Molinillo can be consumed as a normal gem, but inside RubyGems/Bundler, it should merely be considered a part of the host project, not as a dependency.

As usual, @segiddins puts it more thoroughly than I do. @voxik, in the end it boils down to two things:

- We can only guarantee correct functioning with the version that we ship with.
- Bundler and Rubygems can't have dynamically resolved dependencies.

**Comment by voxik** [comment]
> RubyGems and Bundler need to vendor Molinillo because they're unable to have any dynamically resolved dependencies.

And this is not true, since Bundler has one dependency and that is RubyGems which are going to ship Molinillo.

Yes, I understand that Bundler supports older versions of RubyGems, where Molinillo was not available. For that cases, there is rubygems_integration.rb in Bundler and it should look like:

1. Has RubyGems integrated Molinillo? Lets use it.
2. RubyGems has not resolver yet? Lets use our own resolver.

I can't see why Bundler should have its own resolver, when RubyGems has its resolver and by coincidence it is even the same resolver.

**Comment by andremedeiros** [comment]
Well, it's not a dependency in the traditional sense. We support a vast array of versions, and we work under the assumption that if bundler exists, that dependency is met (as in, you have to have ruby gems to run `gem install bundler`.)

Besides, what happens with every single rubygems version up to the point where they got Molinillo? We'd have to provide an alternative anyway.

**Comment by voxik** [comment]
Bundler supports plenty of new functionality of RubyGems where available and provides some fallback implementation in case it is needed for Bundler functionality (see lib/bundler/rubygems_integration.rb). This is not any different.

**Comment by segiddins** [comment]
This is different, since Bundler relies on having very specific versions of Molinillo. We'd still need to vendor it one way or another.

**Comment by voxik** [comment]
@segiddins Sorry, but it is not clear to me what is so specific about version of Molinillo used by Bundler. Could you please elaborate?

**Comment by segiddins** [comment]
Bundler is only tested against a single version of Molinillo, and so far has used features from the latest version each time I've done a release. Additionally, Bundler will have to vendor Molinillo until it's minimum RubyGem's requirement is 2.5, which won't happen for _years_.
